### DISCUSSION TITLE
Print the literal text of an HTML file unrendered

### DISCUSSION THREAD

**Discussion by davb06** [PDConSec/vsc-print#218] [Q&A] [answered]
How can the the full contents of a text file containing HTML be printed, but without any rendering?

This would mirror the printing functionality of Window's Notepad, but enable use of Visual Studio Code's enhanced functionality

As an example, how can all the text below - including HTML tags - be printed, just as it appears, without rendering?
```html
<!doctype html>
<html>
  <head>
    <title>Page title</title>
  </head>
  <body>
    <p>Example paragraph.</p>
  </body>
</html>
``` 

The `Print: Render Markdown` setting has no effect on this.

**Comment by PeterWone** [comment, marked as answer]
No special steps are required. 

This is what happened on my system with default settings and your own sample saved into a file `sample.html`

![image](https://github.com/PDConSec/vsc-print/assets/5498936/b85744a2-2acd-436a-ab15-b7d72398996f)

The renderer that is used will depend on the language identifier determined by VS Code. This is largely determined by the file type extension, so I also tried pasting your HTML into an unsaved document, with this result.

![image](https://github.com/PDConSec/vsc-print/assets/5498936/fe5df630-2bdd-4e7a-b1ad-5694fea68f7b)

> **Aside** To print HTML "without" rendering requires a renderer. If you send the raw HTML to the browser, it will process it just like any other web page. To make it display as source code you have to put it through the source code renderer, which converts angle brackets and whatnot to HTML entities (`&lt;` or `&gt;`) and also does the line numbers and syntax colouring.

I wonder whether VS Code thinks your file contains Markdown. If you embed HTML in Markdown it is rendered, with the exception of processing directives which are not recognised as directives unless they appear before `<html>`. Here's what happens when I save that unsaved HTML and set the file extension to `md`:

![image](https://github.com/PDConSec/vsc-print/assets/5498936/d9582789-11b5-4384-a6ac-70031e65dcd8)

If you're trying to print a snippet of HTML in the middle of a Markdown document because you're writing documentation or an assignment for university, use a fence, just as you did when composing this issue.

````
# Embedded HTML

This is a markdown document. Below is some sample HTML:

```html
<!doctype html>
<html>

<head>
	<title>Page title</title>
</head>

<body>
	<p>Example paragraph.</p>
</body>

</html>
```
````
![image](https://github.com/PDConSec/vsc-print/assets/5498936/ccac0e59-2914-459e-be4b-924b9cbbf1d8)

# Note

If this isn't the behaviour you're seeing, report an issue with details of how you produce the failure and any other information you think might be pertinent.

**Comment by davb06** [comment]
Thanks @PeterWone, your explanation helped me identify why I found this as an issue but you did not.

The issue arises from the language identifier selected by VS Code from the file extension:
>In VS Code, we default the language support for a file based on its filename extension.

- My initial file had a .txt extension, so VS Code defaulted the language as text.
- The example used by @PeterWone  had an .html extension, so the language was defaulted by VS Code as html. (https://code.visualstudio.com/docs/languages/overview#_language-identifier)

Text files are rendered, html files are not rendered.

The solution is to identify to VS Code the files as text.

**Comment by PeterWone** [comment]
Thank you for following up on this. 

You may have identified a bug in the handling of text files. It is necessary to inhibit HTML rendering for plain text, either by wrapping it in a `<PRE>` tag or by escaping `<`, `&` and `>`. I will check whether this is in fact done.

**Comment by PeterWone** [comment]
I checked and the content of plain text files _is_ wrapped in a `<PRE>` block but the browser does not seem to process this as expected. It will be necessary to escape `<&>` characters. I'll raise an issue. Thank you for your help in maintaining the high quality of Print that our users deserve.
